### PR TITLE
[onert] Introduce ExtraTensorIndex

### PR DESCRIPTION
**Pull request by zetwhite**
This PR introduces ExtraTensorIndex.
ExtraTensorIndex will be used to identify extra tensor in tensor registry.

ONE-DCO-1.0-Signed-off-by: seunghui youn <sseung.youn@samsung.com>

draft : https://github.com/Samsung/ONE/pull/13486 
related : https://github.com/Samsung/ONE/issues/13282

### CODE REVIEW

File: runtime/onert/backend/train/ExtraTensorIndex.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+{
+
+template <> struct hash<onert::backend::train::ExtraTensorIndex>
+{
+  size_t operator()(const onert::backend::train::ExtraTensorIndex &index) const noexcept
+  {
+    const auto op_index = index.op_index();
+    const auto sub_index = index.sub_index();
+
+    assert(sizeof(op_index) <= 4);
```
Comment (zetwhite) [inline runtime/onert/backend/train/ExtraTensorIndex.h:75]:
```suggestion
    assert(sizeof(op_index) >= sizeof(uint32_t));
```

**Comment by glistening** [comment]
When I read `ExtraTensorIndex`, it was hard at least for me what `ExtraTensor` does mean.
To figure out the meaning of `ExtranTensor`, I searched and found the answer in #13576.

> ExtraTensor is a tensor that is accessed within one operation layer.
> In other words, the scope of the extra tensor is confined to one specific layer.

It would be better to use some terms in your explanation above.
(e.g. LayerScopeTensor, LayerTemporaryTensor, ...)

**Comment by zetwhite** [comment]
> It would be better to use some terms in your explanation above. (e.g. LayerScopeTensor, LayerTemporaryTensor, ...)

Thansk for comments! I also feel awkward with the name 'Extra'. 
@ys44kim asked me similar a questions(https://github.com/Samsung/ONE/pull/13604#discussion_r1706367894) and it feel hard to explain where the name comes from. 

The name `LayerTemporaryTensor` looks good to me. I will rename it soon.

**Review by jyoungyun — DISMISSED** [review]
LGTM

**Comment by zetwhite** [comment]
> The name LayerTemporaryTensor looks good to me. I will rename it soon. 

About renaming it, I will do it at once, after both this PR and https://github.com/Samsung/ONE/pull/13604 are merged.

**Comment by glistening** [comment]
@zetwhite or who is going to merge this,

Please update a typo in commit message.

ExtraTensorIndex will be used to ~_indentify_~ identify extra tensor in tensor registry.

**Comment by zetwhite** [comment]
> Please update a typo in commit message. 

Ah, Nice catch. I updated. 
Thank you

File: runtime/onert/backend/train/ExtraTensorIndex.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+template <> struct hash<onert::backend::train::ExtraTensorIndex>
+{
+  size_t operator()(const onert::backend::train::ExtraTensorIndex &index) const noexcept
+  {
+    const auto op_index = index.op_index();
+    const auto sub_index = index.sub_index();
+
+    assert(sizeof(op_index) <= sizeof(uint32_t));
+    static_assert(sizeof(size_t) >= sizeof(uint32_t),
+                  "ExtraTensorIndex's hash creation error, size_t size is less than uint32_t");
```
Comment (glistening) [inline runtime/onert/backend/train/ExtraTensorIndex.h:77]: (optional) Personally, I prefer to remove this assert.

In general, sizeof(size_t) is 4 on 32-bit system, 8 on 64-bit system. Both cases, it is >= 4 (= sizeof(uint32_t)). 

If any, I don't think `onert` will support the kind of system.

**Review by ragmani — APPROVED** [review]
LGTM

**Review by jyoungyun — APPROVED** [review]
LGTM

**Pull request merged**